Force GCE metadata querying to be done without proxies

erwanor opened the pull request:
Following issue #114

aaronwinter committed:
Force GCE metadata querying to be done without proxies

Following issue #114

aaronwinter committed:
Disable autodetect proxy handling + adjust to the right library

aaronwinter committed:
Correct build_opener method calling

aaronwinter committed:
Adjust instantiation signature for ProxyHandler

aaronwinter commented:
If my understanding is completing the test is failing because urlopen should be called at least once (see [test_oauth2client.py](https://github.com/google/oauth2client/blob/master/tests/test_oauth2client.py#L217). That being said according to [urllib2.py](http://www.opensource.apple.com/source/python/python-3/python/Lib/urllib2.py) urlopen ends up making a call to open anyway. Is that ok @craigcitro ?

craigcitro commented:
@aaronwinter i suspect it's just that the call doesn't match the expected one in the mock; can you look at updating it? (it's probably just that it needs the args to match.)

craigcitro commented on oauth2client/client.py:
```diff
@@ -938,7 +938,8 @@ def _detect_gce_environment(urlopen=None):
   # "unlikely".
   try:
-    response = urlopen('http://169.254.169.254/', timeout=1)
+    no_proxy_conf = urllib.request.build_opener(urllib.request.ProxyHandler({}))
```
we can also pull this line out of the `try` block.

aaronwinter committed:
Pull OpenerDirector instance out of try block

aaronwinter committed:
Update argument in unit tests to fit recent changes made in GCE environment detection

aaronwinter committed:
Merge pull request #2 from google/master

Use six to call raw_input for Python 3 support

aaronwinter commented:
@craigcitro I have updated the arguments passed to open but the test is still failing. I am not very well versed with unittest. Can you take a quick look at it?

craigcitro commented on tests/test_oauth2client.py:
```diff
@@ -209,21 +209,23 @@ def test_get_environment_gae_local(self):
+    with mock.patch.object(no_proxy_conf, 'open',
                            return_value=response,
-                           autospec=True) as urlopen:
+                           autospec=True) as open:
```
`open` is a dangerous name here (since it overrides the builtin); maybe `no_proxy_open`?

craigcitro commented on tests/test_oauth2client.py:
```diff
@@ -209,21 +209,23 @@ def test_get_environment_gae_local(self):
+                           autospec=True) as open:
       self.assertEqual('GCE_PRODUCTION', _get_environment())
-      urlopen.assert_called_once_with(
+      no_proxy_conf.open.assert_called_once_with(
```
i suspect you just want to use `open` here, since that's your mock. (or, you know, whatever you rename it to based on the above. `;)`)

nathanielmanistaatgoogle commented:
Please squash commits and push a new draft that passes tests?

aaronwinter committed:
Merge pull request #5 from google/master

Squash commits from master repository

aaronwinter committed:
Update the patching name of the no_proxy_conf object to the Mock

Previously the mock's name was overriding the built-in 'open', also the arguments were no matching correctly for both test_get_environment_gce_production and test_get_environment_unknown. This should fix the problem!

dhermes commented:
@nathanielmanistaatgoogle I'm closing this since #293 handles it. @aaronwinter Feel free to re-open if you feel like my change misses something.

ASIDE: It'd be nice to get it in before cutting `1.5.0`.
